reana_workflow_engine_snakemake/executor.py

```diff
@@ -154,36 +143,79 @@ def _get_container_image(self, job: Job) -> str:
             log.info(f"No environment specified, falling back to: {container_image}")
         return container_image
 
-    def _handle_job_status(self, job, status):
+    def _handle_job_status(
+        self, job, job_status: JobStatus, workflow_status: RunStatus
+    ) -> None:
         workflow_uuid = os.getenv("workflow_uuid", "default")
         job_id = job.reana_job_id
-        log.info(f"{status} job: {job_id}")
+        log.info(f"{job_status} job: {job_id}")
         message = {
             "progress": build_progress_message(
-                **{status: {"total": 1, "job_ids": [job_id]}}
+                **{job_status.name: {"total": 1, "job_ids": [job_id]}}
             )
         }
-        status_running = 1
-        status_failed = 3
-        status_mapping = {"finished": status_running, "failed": status_failed}
         self.publisher.publish_workflow_status(
-            workflow_uuid, status_mapping[status], message=message
+            workflow_uuid, workflow_status.value, message=message
         )
 
-    def handle_job_success(self, job):
+    def handle_job_success(self, job) -> None:
         """Override job success method to publish job status."""
         # override handle_touch = True, to enable `touch()` in Snakefiles
         super(ClusterExecutor, self).handle_job_success(
-            job, upload_remote=False, handle_log=False, handle_touch=True
+            job, upload_remote=False, handle_log=False, handle_touch=False
         )
 
-        self._handle_job_status(job, "finished")
+        self._handle_job_status(
+            job, job_status=JobStatus.finished, workflow_status=RunStatus.running
+        )
 
-    def handle_job_error(self, job):
+    def handle_job_error(self, job) -> None:
         """Override job error method to publish job status."""
         super().handle_job_error(job)
 
-        self._handle_job_status(job, "failed")
+        self._handle_job_status(
+            job, job_status=JobStatus.failed, workflow_status=RunStatus.failed
+        )
+
+    def _get_job_status_from_controller(self, job_id: str) -> str:
+        response = self.rjc_api_client.check_status(job_id)
+        return response["status"]
+
+    def _wait_for_jobs(self):
+        """Override _wait_for_jobs method to poll job-controller for job statuses.
+
+        Original _wait_for_jobs method was checking success/failure via .jobfinished or .jobfailed files.
+        """
+        while True:
+            with self.lock:
+                if not self.wait:
+                    return
+                active_jobs = self.active_jobs
+                self.active_jobs = []
+                still_running = []
+
+            for active_job in active_jobs:
+                job_id = active_job.job.reana_job_id
+
+                try:
+                    status = self._get_job_status_from_controller(job_id)
+                except Exception as error:
```

Comment:
Three scenarios I can think of: 

- network failure between engine and r-j-c, unlikely as they are on the same machine; 

- response *404*, handled by r-j-c client so need to catch "HTTPNotFound" exception

 - response *500*, not handled by r-j-c client, it is actually silenced ([line](https://github.com/reanahub/reana-commons/blob/c28e8c2d520bca2bd737c31dc46f2768fcbc00aa/reana_commons/api_client.py#L184)); in this case, the response content will not be correct, so `response["status"]` might throw a key error. But, it can be fixed with `.get("status")`.

Reply:
I tried to make try/catch blocks more specific and clear. I moved this logic to `get_job_status_from_controller`. In case of an error, it will return failed status for a job.

Reply:
*Note:* `rjc_api_client.check_status` method is misleading. Docstring says that method checks the status of a job, but it actually returns the response of the `get_job` API with some additional checks for 404 errors. Just wanted to point that out.